File: algorithms/N/Number of Valid Words in a Sentence/Number of Valid Words in a Sentence.java
class Solution {
    public int countValidWords(String sentence) {
        String regex = "^([a-z]+(-?[a-z]+)?)?(!|\\.|,)?$";
        String r2 = "[^0-9]+";
        String[] arr = sentence.split("\\s+");
        int ans = 0;
        for(String s: arr)
        {
            if(s.matches(regex) && s.matches(r2))
            {
                ans++;
                //System.out.println(s);
            }
        }
        return ans;
    }
}

File: algorithms/N/Number of Valid Words in a Sentence/Number of Valid Words in a Sentence.py
import re
class Solution:
    def countValidWords(self, sentence: str) -> int:
        
        # parse and get each word from sentence
        words = sentence.split()
        
        # regular expression pattern for valid words
        pattern = re.compile( r'^([a-z]+\-?[a-z]+[!\.,]?)$|^([a-z]*[!\.,]?)$' )
        
        # valid word count
        count = 0
        
        # scan each word from word pool
        for word in words:
            
            # judge whether current word is valid or not
            match = re.match(pattern, word)
            
            if match:
                count+=1
                
        return count
